Support AVR for inline asm!

couchand opened the pull request:
A first pass at support for the AVR platform in inline `asm!`.  Passes the initial compiler tests, have not yet done more complete verification.

In particular, the register classes could use a lot more fleshing out, this draft PR so far only includes the most basic.

cc @Amanieu @dylanmckay

rust-highfive commented:
r? @wesleywiser

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4326062704?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ca568512b2d78b9d5880f7570a972cf9ae89d8d9/checks/4326062704/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking gccjit_sys v0.0.1 (https://github.com/antoyo/gccjit.rs#2d4fea73)
    Checking gccjit v1.0.0 (https://github.com/antoyo/gccjit.rs#2d4fea73)
    Checking object v0.25.3
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0004]: non-exhaustive patterns: `Avr(_)` not covered
    |
    |
558 |         InlineAsmRegOrRegClass::RegClass(reg) => match reg {
    |                                                        ^^^ pattern `Avr(_)` not covered
   ::: /checkout/compiler/rustc_target/src/asm/mod.rs:407:5
    |
    |
407 |     Avr(AvrInlineAsmRegClass),
    |
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = note: the matched value is of type `rustc_target::asm::InlineAsmRegClass`

error[E0004]: non-exhaustive patterns: `Avr(_)` not covered
    |
616 |     match reg {
616 |     match reg {
    |           ^^^ pattern `Avr(_)` not covered
   ::: /checkout/compiler/rustc_target/src/asm/mod.rs:407:5
    |
    |
407 |     Avr(AvrInlineAsmRegClass),
    |
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = note: the matched value is of type `rustc_target::asm::InlineAsmRegClass`

error[E0004]: non-exhaustive patterns: `Avr(_)` not covered
    |
724 |     match reg {
724 |     match reg {
    |           ^^^ pattern `Avr(_)` not covered
   ::: /checkout/compiler/rustc_target/src/asm/mod.rs:407:5
    |
    |
407 |     Avr(AvrInlineAsmRegClass),
    |
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
    = note: the matched value is of type `rustc_target::asm::InlineAsmRegClass`
For more information about this error, try `rustc --explain E0004`.
error: could not compile `rustc_codegen_gcc` due to 3 previous errors
Build completed unsuccessfully in 0:03:37
```

</details>

Amanieu commented on compiler/rustc_target/src/asm/avr.rs:
```diff
@@ -0,0 +1,128 @@
+        r31: reg = ["r31", "ZH"],
+
+        X: reg_xyz = ["r27r26", "X"],
+        Y: reg_xyz = ["r29r28", "Y"],
```
If you have a look at `lib/Target/AVR/AVRRegisterInfo.cpp` in the LLVM source you will see that `Y` is actually reserved for the stack pointer. Also `r0` and `r1` are reserved.

couchand replied:
It looks like for `Y` the way to go is adding the `#error` pragma here to keep the user's hands off of it.

I think we'd need some way to name `r0` and `r1`.  For instance, to get the result of a `MUL`:

```
let a = 1u8;
let b = 1u8;
let res: u16;
asm!(
    "mul {}, {}",
    in(reg) a,
    in(reg) b,
    lateout("r1r0") res,
);
```

Is there another architecture that I can look to for an example of how to handle this?

---

(and just to clarify because it confused me for a moment, `Y` is reserved for the frame pointer, the stack pointer has a dedicated register `SPH:SPL`)

Amanieu replied:
> I think we'd need some way to name r0 and r1.

Or just leave it to the user to reassemble the value from `r0` and `r1` themselves. In the end it should be optimized the same way by the compiler.

couchand replied:
Sorry, I wasn't clear.  If `r0` and `r1` are reserved, and thus probably shouldn't be nameable for inline `asm!`, how is the user expected to get the results from them?

Or to put it another way, what should I do with the information that `lib/Target/AVR/AVRRegisterInfo.cpp` reserves these registers if not exclude them from use?

dylanmckay replied:
> Sorry, I wasn't clear. If r0 and r1 are reserved, and thus probably shouldn't be nameable for inline asm!, how is the user expected to get the results from them?

Both registers `r1` and `r0` should be accessible. Any author of some piece Rust code targeting AVR should be able to name these registers in inline ASM.

The zero register should always be accessible even in inline asm so that the dev can pass `0` to an instruction which only supports registers as inputs, not immediates.

See this avr-gcc wiki doc https://gcc.gnu.org/wiki/avr-gcc#Fixed_Registers which shows that both registers should be available in inline asm.

It is a little bit finicky with LLVM instruction reordering, but it's always like that and I don't see a problem with your approach. You won't be able to say, run a piece of C code (i.e. something that executes `MUL`) and then use inline ASM to reassemble `r1`/`r0` because LLVM inline assembler does not parse/inspect into the raw assembly text (only the constraints/list), so LLVM won't know you're referring to r1/r0 and so will not be able to preserve it during transformation. You would be able to use inline ASM to reassemble r1/r0 from the result of an instruction that is also inside the inline asm block, because the internal contents of any inline asm block would not be reordered. **In summary, in this paragraph all I am trying to say that it is perfectly acceptable and allowable to use `r1`/`r0` in inline ASM, but you cannot rely on their values across inline asm block boundaries**

couchand replied:
It sounds like there should be a register class just for `r1`/`r0`, they shouldn't be in `reg` because we wouldn't want to pick either of them for "any old register".

> The zero register should always be accessible

Thanks for bringing this up.  I'll need to add to the documentation that if a programmer writes to `r1` in any way that they must restore it to zero before the end of the `asm!` block.

> it is perfectly acceptable and allowable to use r1/r0 in inline ASM, but you cannot rely on their values across inline asm block boundaries.

Does that mean that you shouldn't be able to use them as inputs or outputs to the block?  For instance, would my code sample up thread be expected to work, or does that count as "across block boundaries"?

Amanieu replied:
I agree that `r0` and `r1` *should* be usable as operands in `asm!`. However the reality is that LLVM marks these as reserved registers, which causes LLVM's register allocator to ignore these and assume they are never touched by inline assembly. If you try to actually use `r0`/`r1` as operands then things will break in unpredictable ways, which is why we always prevent reserved registers from being used in `asm!`.

The only "correct" way to use `r0` and `r1` in `asm!` is to save their value to another register (which should be properly marked as clobbered) and then restoring these values at the end of the assembly snippet. This is unavoidable until the LLVM backend is fixed to no longer treat these registers as reserved.

Amanieu replied:
Source: https://github.com/llvm/llvm-project/blob/27c9e8b45b25614a92539ac6787dbb5670d950b3/llvm/lib/Target/AVR/AVRRegisterInfo.cpp#L53

couchand replied:
I am having a hard time understanding exactly what LLVM means by reserved.  The only authority I've found is the comment above `getReservedRegs` here: https://github.com/llvm/llvm-project/blob/e356027016c6365b3d8924f54c33e2c63d931492/llvm/include/llvm/CodeGen/TargetRegisterInfo.h#L524-L527

```
  /// A reserved register:
  /// - is not allocatable
  /// - is considered always live
  /// - is ignored by liveness tracking
```

I don't think I know enough to tell if that should be applied to `r0` and `r1` or not.  I'm inclined to trust your judgement there @Amanieu.

> This is unavoidable until the LLVM backend is fixed to no longer treat these registers as reserved.

If this is where we're going, I'm happy to submit this change to LLVM.  @dylanmckay, since I suspect you'll be reviewing the differential, does that change make sense to you?

In the meantime, I'll make `r0` and `r1` error out here with a helpful message.

dylanmckay replied:
> > This is unavoidable until the LLVM backend is fixed to no longer treat these registers as reserved.

> If this is where we're going, I'm happy to submit this change to LLVM. @dylanmckay, since I suspect you'll be reviewing the differential, does that change make sense to you?

From my point of view, if AVR-GCC compatibility is kept, then the change is definitely good. If not, maybe it's possible but certainly more complicated :+1:

Amanieu commented on src/doc/unstable-book/src/library-features/asm.md:
```diff
@@ -593,6 +595,7 @@ Here is the list of currently supported register classes:
 | wasm32 | `local` | None\* | `r` |
 | BPF | `reg` | `r[0-10]` | `r` |
 | BPF | `wreg` | `w[0-10]` | `w` |
+| AVR | `reg` | `r[0-31]`, `xh`, `xl`, `yh`, `yl`, `zh`, `zl` | `r` |
```
`reg_xyz` is missing from the documentation.

couchand replied:
Will resolve once the various classes are ironed out.

Amanieu commented on compiler/rustc_target/src/asm/avr.rs:
```diff
@@ -0,0 +1,128 @@
+        r28: reg = ["r28", "YL"],
+        r29: reg = ["r29", "YH"],
+        r30: reg = ["r30", "ZL"],
+        r31: reg = ["r31", "ZH"],
```
On all other architectures the register names are lowercase. Is this expected to be uppercase here? The compiler is case-sensitive.

couchand replied:
The "special" registers (`[XYZ][LH]?` and `SP[LH]?` as well as some more exotic "registers") seem to generally be written in uppercase in docs & example code.  But perhaps for consistency it would be preferable to write them lowercase here?  I'm happy to switch that over.

dylanmckay replied:
I vote to keep `X,Y,Z` registers and other registers mentioned above by Andrew as uppercase for consistency with Atmel docs. The GNU and LLVM assemblers for AVR accepts either case. If it were lowercase-only I think it would surprise users (if it's case sensitive, we should at least put it in the case the're most likely going to use - which my best guess is upper case but who knows with production code)

couchand replied:
I agree that the copy-paste from the Atmel docs is a key use-case, and so should just work.

We could make both uppercase and lowercase aliases, so that it behaves the same as the assemblers.  But something about that idea seems a bit ugly.

What do you think @Amanieu?

Amanieu replied:
I'm not too familiar with AVR so I will leave the decision to you. Just make sure the documentation in the unstable book is consistent with what is actually implemented.

Amanieu commented on src/doc/unstable-book/src/library-features/asm.md:
```diff
@@ -708,6 +712,12 @@ Some registers have multiple names. These are all treated by the compiler as ide
+| AVR | `yh` | `r29` |
+| AVR | `yl` | `r28` |
+| AVR | `zh` | `r31` |
+| AVR | `zl` | `r30` |
```
These are uppercase in the code.

couchand replied:
Will address once the capitalization question is answered.

Amanieu commented on src/test/assembly/asm/avr-types.rs:
```diff
@@ -0,0 +1,134 @@
+// CHECK: ;APP
+// CHECK: mov r27, r27
+// CHECK: ;NO_APP
+check_reg!(xh_i8 i8 "XH");
```
Tests for `reg_xyz` are missing.

couchand replied:
Resolved.

Amanieu commented on src/test/assembly/asm/avr-types.rs:
```diff
@@ -0,0 +1,134 @@
+#[no_mangle]
+pub unsafe fn sym_fn() {
+    asm!("call {}", sym extern_func);
+}
```
Missing tests for `sym_static`.

Amanieu commented on compiler/rustc_codegen_llvm/src/asm.rs:
```diff
@@ -642,6 +643,8 @@ fn reg_to_llvm(reg: InlineAsmRegOrRegClass, layout: Option<&TyAndLayout<'tcx>>)
             InlineAsmRegClass::Bpf(BpfInlineAsmRegClass::reg) => "r",
             InlineAsmRegClass::Bpf(BpfInlineAsmRegClass::wreg) => "w",
+            InlineAsmRegClass::Avr(AvrInlineAsmRegClass::reg) => "r",
+            InlineAsmRegClass::Avr(AvrInlineAsmRegClass::reg_xyz) => "r",
```
Shouldn't this be `e`? According to the constraint list here: https://www.nongnu.org/avr-libc/user-manual/inline_asm.html

couchand replied:
Indeed!  Addressed.

Amanieu reviewed (commented):
Great work! Regarding register classes, just make sure you have enough to accurate describe the constraints that would be needed for various instructions. See https://www.nongnu.org/avr-libc/user-manual/inline_asm.html for a mostly complete list of constraint code used by avr-gcc and LLVM. For example, I can see some instructions which only accept a "high" register (but then again I'm not very familiar with AVR assembly).

wesleywiser commented:
r? @Amanieu

couchand commented:
Thanks for your review @Amanieu!  I've replied in the places where I have questions.

Is there any style guidance on the translation from constraint codes/descriptions to register class names?  Here is my proposal for a minimal set of constraint codes to support:

- `r` for any 8-bit register
- `w` for any 16-bit register pair
- `d` for an 8-bit upper register
- `e` for any pointer register pair (just `X` and `Z` I guess)
- `z` for the `Z` register (and also any base pointer register since `Y` is not usable) [does this even need to be a class?]

The first two could possibly be named `reg` and `wreg` to reuse symbols from other architectures, but I'm not sure what the others should be named.

couchand commented:
I've pushed up a variety of changes that addresses most (but not all) of the outstanding items.  Please take a look when you have availability @Amanieu.  Thanks!

couchand commented:
I think I'm not entirely clear on how I should be handling the register pairs.  Any pair can be used as an argument to the 16-bit MOVW instruction, and the upper pairs can be an argument to the 16-bit immediate instructions like ANDIW.

Also, each of the pointer pairs, which have special behavior for the indirect memory operations, can be used as two general-purpose 8-bit registers when passed to any other instruction.

The 8-bit and 16-bit operation expect the lower register of the pair to be named, with the upper register implicit.

Are the changes I've made here consistent with that usage?

dylanmckay commented on compiler/rustc_codegen_llvm/src/asm.rs:
```diff
@@ -642,6 +643,11 @@ fn reg_to_llvm(reg: InlineAsmRegOrRegClass, layout: Option<&TyAndLayout<'tcx>>)
+            InlineAsmRegClass::Avr(AvrInlineAsmRegClass::reg) => "r",
+            InlineAsmRegClass::Avr(AvrInlineAsmRegClass::reg_upper) => "d",
+            InlineAsmRegClass::Avr(AvrInlineAsmRegClass::reg_pair) => "r",
+            InlineAsmRegClass::Avr(AvrInlineAsmRegClass::wreg) => "w",
```
The AVR LLVM backend knows 3 special pointer registers X, Y, Z (r25-r31 or so). This register class forms the "pointer register class"

I haven't seen the `W` register (I'm inferring it would be `r25r24`) mentioned in the instruction set docs anywhere. Obviously r25r24 still exists but I don't think it is conventional to give it the name `W` because the instruction set does not allow it to be used as an pointer so it's not really a special register deserving of an alphabetic alias.

Recommend removing `wreg` from AVR inline asm. Unless there's a reason/usecase I'm missing

dylanmckay replied:
Ignore this comment, I just read the manual (properly this time) and wreg is an existing constraint and you have implemented the correct registers in it

dylanmckay commented:
> I think I'm not entirely clear on how I should be handling the register pairs. Any pair can be used as an argument to the 16-bit MOVW instruction, and the upper pairs can be an argument to the 16-bit immediate instructions like ADIW.
> 
> Also, each of the pointer pairs, which have special behavior for the indirect memory operations, can be used as two general-purpose 8-bit registers when passed to any other instruction.
> 
> The 8-bit and 16-bit operation expect the lower register of the pair to be named, with the upper register implicit.
> 
> Are the changes I've made here consistent with that usage?

One point I'll make is that the instruction set is pretty ad-hoc/bolted on parts in ways, with even the registers supported for an "instruction" being target MCU specific. You will see instructions like: `LPM`: only ever works on the Z register, therefore, to use in inline ASM there needs to be a register class that can only ever resolve to the Z register[0] so GCC has an inline asm reg class constraint named `z`.

Another instruction complicating it is `STD/LDD` (store/load to memory with pointer offset/displacement). On some obscure AVR chips, they are only defined for Y and Z pointer registers and not the X pointer register (but you can probably ignore this case because the LLVM backend doesn't even make this distinction). But if the backend did make the distinction in the future as it should already be doing, then for the Rust code to work correct across AVR chips, the Rust code would need to explicitly opt-in to one of the `x, y, z` reg class constraints (redundantly excluding registers on the high end device that could be used, from being used), or the Rust compiler itself would need to expose an inline asm reg class constraint whose included registers were target device specific so that the Rust could could just use that and have it just work.

Suffice to say, I think it's best to land a core set like you're doing and add the others later. It's probably worth adding 3 more extra register classes in this PR (`x`, `y`, `z`, that can only resolve to their own respective implied registers) which will allow Rust code to at least use instructions like `LPM` and also work around any chip-specific/subtarget feature issues.

- [0] Inline asm list: https://www.nongnu.org/avr-libc/user-manual/inline_asm.html

couchand commented:
Thanks for the review @dylanmckay!

The point is well received that every AVR device is a new beast.  For myself, I'm usually working on ones that don't even have a hardware `MUL` (so that conversation is all a bit theoretical!).  A very conservative use of `asm!` might have a good chance of being portable across cores, but anything very complicated will need special attention paid to the specific core(s) in use.  I think that's a reasonable trade off to make.

> It's probably worth adding 3 more extra register classes in this PR (x, y, z, that can only resolve to their own respective implied registers) which will allow Rust code to at least use instructions like LPM and also work around any chip-specific/subtarget feature issues.

Are all the use-cases for these handled by the specific-register syntax of `asm!`?  That is, is it sufficient that the user can write

```rust
let mut byte = MaybeUninit::uninit();
let input = 42u8;
asm!(
    "st X, {}",
    inout("X") byte.as_mut_ptr(),
    in(reg) input,
);
```

or should you also be able to say

```rust
let mut byte = MaybeUninit::uninit();
let input = 42u8;
asm!(
    "st {}, {}",
    inout(x) byte.as_mut_ptr(),
    in(reg) input,
);
```

Amanieu commented:
I would recommend adding template modifiers for register pairs to print just the high/low register of a pair. Have a look at how it's done for ARM where the `e` and `f` modifiers allow printing the low/high d-reg component of a q-reg. In the AVR backend the equivalent modifiers are `A` and `B`.

Amanieu commented on compiler/rustc_target/src/asm/avr.rs:
```diff
@@ -36,15 +39,20 @@ impl AvrInlineAsmRegClass {
+            Self::reg_upper => types! { _: I8; },
+            Self::reg_pair => types! { _: I16; },
+            Self::wreg => types! { _: I16; },
+            Self::ptr => types! { _: I16; },
```
This should probably have `reg` somewhere in the name for consistency with the others. `reg_ptr`? `ptr_reg`?

couchand replied:
Not to mention how strange a bare `ptr` looks...  Will update to `reg_ptr`.

Amanieu commented on compiler/rustc_target/src/asm/avr.rs:
```diff
@@ -89,7 +114,24 @@ impl AvrInlineAsmReg {
+            r7r6 => "r6",
+            r5r4 => "r4",
+            r3r2 => "r2",
+            r1r0 => "r0",
```
This function should emit the *exact* same string that LLVM's inline assembler will for the given register. It's not currently used by the backend, but will eventually be used for the fallback assembler where we generate a .S file and pass it to an external assembly (e.g. for the cranelift backend which doesn't support assembly natively).

couchand replied:
Understood.  I believe that would come from the `AsmName` field of the `llvm/Target/Register` class, so in this case it should be `r1:r0` (with the colon).  Will update.

couchand commented:
> I would recommend adding template modifiers for register pairs to print just the high/low register of a pair. Have a look at how it's done for ARM where the e and f modifiers allow printing the low/high d-reg component of a q-reg. In the AVR backend the equivalent modifiers are A and B.

That's a great idea.  I'll go ahead and add those.

couchand commented:
Thank you @Amanieu and @dylanmckay for your continued reviews and many thoughtful comments.  I have incorporated all of the feedback thus far and rebased this branch.  Please take another look when you have an opportunity.

Amanieu commented on compiler/rustc_codegen_llvm/src/asm.rs:
```diff
@@ -322,6 +322,7 @@ impl AsmBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 InlineAsmArch::SpirV => {}
                 InlineAsmArch::Wasm32 | InlineAsmArch::Wasm64 => {}
                 InlineAsmArch::Bpf => {}
+                InlineAsmArch::Avr => {}
```
I had to look this up in LLVM's internals, but the clobber for the status register is `"~{sreg}"`, which should be marked as clobbered when `preserves_flags` is not set. Also the documentation in the unstable book should be updated to reflect this.

couchand replied:
I've added that.

So many of the other platforms have an empty branch here it was easy to not spend much thought on this clause. But the realization that I missed this made be worried there might be more.  I just went through again and checked for other copy-pasta, didn't notice any concerns.

The one other place we have a null implementation that could potentially have code is `suggest_modifier` -- that one I did consider if there was a heuristic to use to know which byte you mean.  But in the end it seems equally likely you'd want either byte, and there doesn't seem to be a way to suggest *two* modifiers.

Amanieu commented on src/doc/unstable-book/src/library-features/asm.md:
```diff
@@ -732,6 +745,7 @@ Some registers cannot be used for input or output operands:
+| AVR | `r0`, `r1`, `r1r0` | Due to an issue in LLVM, the `r0` and `r1` registers cannot be used as inputs or outputs.  If modified, they must be restored to their original values before the end of the block. |
 
 In some cases LLVM will allocate a "reserved register" for `reg` operands even though this register cannot be explicitly specified. Assembly code making use of reserved registers should be careful since `reg` operands may alias with those registers. Reserved registers are the frame pointer and base pointer
 - The frame pointer and LLVM base pointer on all architectures.
```
Add `r0` and `r1` to the list of reserved registers here.

couchand replied:
Added.

Amanieu commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit c6e8ae1a6ceeffe343c214859e4eb686b7a71c95 has been approved by `Amanieu`

<!-- @bors r=Amanieu c6e8ae1a6ceeffe343c214859e4eb686b7a71c95 -->
<!-- homu: {"type":"Approved","sha":"c6e8ae1a6ceeffe343c214859e4eb686b7a71c95","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit c6e8ae1a6ceeffe343c214859e4eb686b7a71c95 with merge 0b6f079e4987ded15c13a15b734e7cfb8176839f...
<!-- homu: {"type":"BuildStarted","head_sha":"c6e8ae1a6ceeffe343c214859e4eb686b7a71c95","merge_sha":"0b6f079e4987ded15c13a15b734e7cfb8176839f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4447335042?check_suite_focus=true)
Approved by: Amanieu
Pushing 0b6f079e4987ded15c13a15b734e7cfb8176839f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Amanieu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4447335042?check_suite_focus=true"},"merge_sha":"0b6f079e4987ded15c13a15b734e7cfb8176839f"} -->

rust-timer commented:
Finished benchmarking commit (0b6f079e4987ded15c13a15b734e7cfb8176839f): [comparison url](https://perf.rust-lang.org/compare.html?start=c5c94945096265b6d675b2f55a247c4799de8d87&end=0b6f079e4987ded15c13a15b734e7cfb8176839f).

**Summary**: This change led to moderate relevant regressions 😿 in compiler performance.
- Moderate regression in instruction counts (up to 0.8% on `incr-unchanged` builds of `externs`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression

couchand commented:
I have no idea what needs to be done to investigate this performance issue.  From the linked results I can see there are five benchmark/profile/scenario rows with issues, and four of them are the externs benchmark.

But I don't know what to do next. What are these externs benchmarks?  Why would inline assembler changes for a platform that presumably isn't exercised by the benchmark make such a significant change?  What's the usual range on these results?

What does "sufficient written justification" require?  Are there any examples or documentation that explain what's needed?

Amanieu commented:
The performance issue is almost certainly not related to this PR.

couchand commented:
> The performance issue is almost certainly not related to this PR.

Good to hear. Is there anything actionable for me here, then?

pnkfelix commented:
@couchand no,  I don't think there's anything actionable for you here. If anything, the performance team might want to double check what's going on with that benchmark, and/or our noise suppression techniques.

pnkfelix commented:
I think we can ignore this regression. `externs` is a pathological benchmark (3000 `extern` function declarations) and I do not think we should worry about it going up by 0.8%, especially since a different like that could easily be due to memory alignment effects in a benchmark like this.

@rustbot label: +perf-regression-triaged
